README.md
<img src="https://raw.githubusercontent.com/magic4dev/systemdy/master/systemdy.png" width="1012px" alt="Systemdy Logo">

![Gem](https://img.shields.io/gem/v/systemdy?style=flat-square)
![Gem](https://img.shields.io/gem/dt/systemdy?style=flat-square)
[![GitHub forks](https://img.shields.io/github/forks/magic4dev/systemdy?style=flat-square)](https://github.com/magic4dev/systemdy/network)
[![GitHub issues](https://img.shields.io/github/issues/magic4dev/systemdy?style=flat-square)](https://github.com/magic4dev/systemdy/issues)
[![GitHub license](https://img.shields.io/github/license/magic4dev/systemdy?style=flat-square)](https://github.com/magic4dev/systemdy/blob/master/LICENSE.txt)
[![GitHub stars](https://img.shields.io/github/stars/magic4dev/systemdy?style=flat-square)](https://github.com/magic4dev/systemdy/stargazers)

# Systemdy

A lightweight gem for interact with Systemd.

If your goal is to develop software to quickly manage systemd services or journalctl logs, this gem is for you! :grin:


## Features

- Lightweight
- Expressive classes and methods
- Manage Systemd's services lifecycle with minimal effort!
- Extract and manipulate Journalctl logs with a single action!

## Table of contents
* [Installation](#installation)
* [Dependencies](#dependencies)
* [Usage](#usage)
* [Manage services](#manage-services)
  * [Create a Systemdy Service object for control a service](#create-a-systemdy-service-object-for-control-a-service)
  * [Check if the provided service exist](#check-if-the-provided-service-exist)
  * [Check if the provided service is enabled](#check-if-the-provided-service-is-enabled)
  * [Check if the provided service is active](#check-if-the-provided-service-is-active)
  * [Check the current status of the provided service](#check-the-current-status-of-the-provided-service)
  * [Display all the properties for the provided service](#display-all-the-properties-for-the-provided-service)
  * [Get service port](#get-service-port)
  * [Get service protocol](#get-service-protocol)
  * [Start the service](#start-the-service)
  * [Restart the service](#restart-the-service)
  * [Stop the service](#stop-the-service)
  * [Enable the service](#enable-the-service)
  * [Disable the service](#disable-the-service)
  * [Reload the service](#reload-the-service)
  * [Mask the service](#mask-the-service)
  * [Unmask the service](#unmask-the-service)
  * [Bonus tip: execute systemctl commands without prompt the password](#bonus-tip-execute-systemctl-commands-without-prompt-the-password)
* [Manage Journalctl logs](#manage-journalctl-logs)
  * [Display kernel logs](#display-kernel-logs)
  * [Display boot logs](#display-boot-logs)
  * [Display unit logs](#display-unit-logs)
  * [Display group_id logs (GUID)](#display-group_id-logs-guid)
  * [Display user_id logs (UID)](#display-user_id-logs-uid)
* [Contributing](#contributing)
  * [Develop a new feature](#develop-a-new-feature) 
  * [Fix a bug](#fix-a-bug) 
* [Note for testing](#note-for-testing) 
* [License](#license)
* [Useful links and resources](#useful-links-and-resources)
* [Acknowledgements](#acknowledgements) 

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'systemdy'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install systemdy

## Dependencies

The only dependecy you need is [systemd](http://www.freedesktop.org/wiki/Software/systemdy/) installed on your system (specifically libsystemd or the older libsystemd-journal) in order to use the gem. Currently the gem support systemd 249 or higher.

## Usage

After installing the gem, the first step is to require it:

```ruby
require 'systemdy'
```

## Manage services

The first goal of this gem is to manage a systemd's service with minimal effort.

This section provides an overview of the Systemdy::Service class for managing the life cycle of a systemd's service.

### Create a Systemdy Service object for control a service

The first step is to create a new instance of the Systemdy::Service class for control the desired service

```ruby
my_postgresql_service = Systemdy::Service.new('postgresql')
```

### Check if the provided service exist

Once our object has been instantiated, you can check if the service is installed on your system

```ruby
my_postgresql_service.exist?
```

For check if a service is installed on your system without create a new instance of the Systemdy::Service class

```ruby
Systemdy::Utility::Validator.check_if_a_service_exist('postgresql')
```
if the provided service is installed on your system this methods return _**true**_, otherwise return _**false**_.

### Check if the provided service is enabled

For check if a service is enabled

```ruby
my_postgresql_service.is_enabled?
```
if the provided service is enabled this method return _**true**_, otherwise return _**false**_.

### Check if the provided service is active

For check if a service is active

```ruby
my_postgresql_service.is_active?
```
if the provided service is active this method return _**true**_, otherwise return _**false**_.

### Check the current status of the provided service

For check the current status of the service 
```ruby
my_postgresql_service.status
```
Once executed, this method return an hash with all the essential service's information like this:

```ruby
 { "Id"=>"postgresql.service",              
   "Description"=>"PostgreSQL RDBMS",          
   "ExecMainPID"=>"48615",
   "LoadState"=>"loaded",
   "ActiveState"=>"active",
   "FragmentPath"=>"/lib/systemd/system/postgresql.service",
   "ActiveEnterTimestamp"=>"Thu 2022-09-29 17:13:07 CEST",
   "InactiveEnterTimestamp"=>"Thu 2022-09-29 17:12:44 CEST",
   "ActiveExitTimestamp"=>"Thu 2022-09-29 17:12:44 CEST",
   "InactiveExitTimestamp"=>"Thu 2022-09-29 17:13:07 CEST"
 } 
```

### Display all the properties for the provided service

For check all the properties of the service 

```ruby
my_postgresql_service.properties
```
Once executed, this method return an hash with all the service's properties like this:

```ruby
 { "Type"=>"oneshot",
    "Restart"=>"no",
    "NotifyAccess"=>"none",
    ....
 }
```

To extract the value of a specific property

```ruby
my_postgresql_service.properties["MemorySwapMax"]
```

### Get service port

For retrieve a service port

```ruby
my_postgresql_service.port
```
Once executed, this method return the provided service port:

```ruby
 "5432"
```

if the provided service has no available port, this method return a message like this:

```ruby
 "[name_of_the_service].service has no port available"
```

### Get service protocol

For retrieve a service protocol

```ruby
my_postgresql_service.protocol
```
Once executed, this method return the provided service protocol:

```ruby
  "tcp"
```

if the provided service has no available protocol, this method return a message like this:

```ruby
 "[name_of_the_service].service has no protocol available"
```

### Start the service

For start the service 
```ruby
my_postgresql_service.start
```

### Restart the service

For restart the service 
```ruby
my_postgresql_service.restart
```

### Stop the service

For stop the service 
```ruby
my_postgresql_service.stop
```

### Enable the service

For enable the service 
```ruby
my_postgresql_service.enable
```

### Disable the service

For disable the service 
```ruby
my_postgresql_service.disable
```

### Reload the service

For reload the service 
```ruby
my_postgresql_service.reload
```

### Mask the service

For mask the service 
```ruby
my_postgresql_service.mask
```

### Unmask the service

For unmask the service 
```ruby
my_postgresql_service.unmask
```

## Bonus tip: execute systemctl commands without prompt the password

The methods **start**, **restart**, **stop**, **enable**, **disable**, **reload**, **mask** and **unmask** detect non-root users and automatically execute the methods with sudo.

By default, sudo needs that a non-root user(in this case you) is authenticated using a password for complete the method execution. This is a great thing, but some times you may need to execute this methods without type the password. 

In this case you need to configure sudo without a password.

The first step for achieve this is gain root access:

```console
non_root_user@my-machine:~$ sudo -i
[sudo] password for non_root_user: 
```

After provided the correct password backup your /etc/sudoers file by typing the following command:

```console
root@my-machine:~# cp /etc/sudoers /root/sudoers.bak 
```

Next step is edit the /etc/sudoers file by typing the visudo command:

```console
root@my-machine:~# visudo 
```

Append/edit the following lines in the /etc/sudoers file and substitute 'your_account_name' with your linux account name.
```console
'your_account_name' ALL=(ALL) NOPASSWD: /bin/systemctl
```

This line anable you to run ‘systemctl’ commands without a password. 

Save with CTRL + X

Type 'exit' on your terminal as follows:

```console
root@my-machine:~# exit
```
Exit from root session and you are ready to go! :sunglasses:

## Manage Journalctl logs

The second goal of this gem is to provide a set of useful methods for retrieving journalctl log data.

This section provides a brief overview of the Systemdy::Journal class for managing journalctl log information.

### Display kernel logs

Sometimes we need to check if the kernel is working properly or if something went wrong to get useful information to help us fix the problem.

For getting the kernel logs 

```ruby
Systemdy::Journal.display_kernel_logs
```
This class method executed with no arguments return an array with: 
* a default size of 10 (the last 10 lines of log)
* a start period not older than '00:00 AM' o' clock (today)
* an end period not newer than the method execution timenow

Obviously this method allows you to filter the kernel logs by passing **3** positional arguments:

* **since**: a start period not older than the specified date

* **to**: an end period not older than the specified date

* **lines**: the exact number of last log lines 

For example:

If you want to analyze the last 50 log lines ranging from a month ago to 10:00 of the current day:

```ruby
Systemdy::Journal.display_kernel_logs(since: '1 month ago', to: '10:00', lines: 50)
```

If you want to analyze the last 20 log lines ranging from a week ago to yesterday:

```ruby
Systemdy::Journal.display_kernel_logs(since: '1 week ago', to: 'yesterday', lines: 20)
```

You can also filter the logs by specific dates in the format YYYY-MM-DD for example:

```ruby
Systemdy::Journal.display_kernel_logs(since: '2022-08-27', lines: 200)
```

If the passed arguments not match anything the method return an array with a message like this:

```ruby
["-- No entries --"]
```

### Display boot logs

Sometimes we need to know if our system is working properly at boot time and if something went wrong to get useful information to help us fix the problem.

For getting the boot logs 

```ruby
Systemdy::Journal.display_boot_logs
```
This class method executed with no arguments return an array with: 
* a default size of 10 (the last 10 lines of log)
* a start period not older than '00:00 AM' o' clock (today)
* an end period not newer than the method execution timenow

Obviously as for the previous method, you can filter the boot logs by passing **4** positional arguments:

* **argument**: in this case the boot's number

* **since**: a start period not older than the specified date

* **to**: an end period not older than the specified date

* **lines**: the exact number of last log lines 

For example:

If you want to analyze the last 50 log lines of the second boot ranging from a month ago to 10:00 of the current day:

```ruby
Systemdy::Journal.display_boot_logs(argument: 2, since: '1 month ago', to: '10:00', lines: 50)
```

If you want to analyze the last 20 log of the third boot lines ranging from a week ago to yesterday:

```ruby
Systemdy::Journal.display_boot_logs(argument: 3, since: '1 week ago', to: 'yesterday', lines: 20)
```

You can also filter the logs by specific dates in the format YYYY-MM-DD for example:

```ruby
Systemdy::Journal.display_boot_logs(since: '2022-08-27', lines: 200)
```
If the passed arguments not match anything the method return an array with a message like this:

```ruby
["-- No entries --"]
```

### Display unit logs

Another common task is to filter logs based on the unit that we need information on. 

To do this, we just simply need to pass the name of the unit as argument, for example we need to analyze postgresql unit logs:

```ruby
Systemdy::Journal.display_unit_logs(argument: 'postgresql')
```

In this case the class method return an array with: 
* a default size of 10 (the last 10 lines of log)
* a start period not older than '00:00 AM' o' clock (today)
* an end period not newer than the method execution timenow

Obviously this method require the unit's name as argument, if executed with no arguments return a message like this:

```ruby
"display_unit_logs require an argument!"
```

As for the previous method, you can filter the unit logs by passing **4** positional arguments:

* **argument**: in this case the unit's name

* **since**: a start period not older than the specified date

* **to**: an end period not older than the specified date

* **lines**: the exact number of last log lines 

For example:

If you want to analyze the last 50 log lines of the potsgresql service ranging from a month ago to 10:00 of the current day:

```ruby
Systemdy::Journal.display_unit_logs(argument: 'postgresql', since: '1 month ago', to: '10:00', lines: 50)
```

If you want to analyze the last 20 log lines of the potsgresql service ranging from a week ago to yesterday:

```ruby
Systemdy::Journal.display_unit_logs(argument: 'postgresql', since: '1 week ago', to: 'yesterday', lines: 20)
```

You can also filter the logs by specific dates in the format YYYY-MM-DD for example:

```ruby
Systemdy::Journal.display_unit_logs(argument: 'postgresql', since: '2022-08-27', lines: 200)
```

If the passed arguments not match anything the method return an array with a message like this:

```ruby
["-- No entries --"]
```

### Display group_id logs (GUID)

To find all messages related to a particular group, we just simply need to pass the GUID of the group as argument, for example:

```ruby
Systemdy::Journal.display_group_id_logs(argument: 1000)
```

In this case the class method return an array with: 
* a default size of 10 (the last 10 lines of log)
* a start period not older than '00:00 AM' o' clock (today)
* an end period not newer than the method execution timenow

Obviously this method require the GUID as argument, if executed with no arguments return a message like this:

```ruby
"display_group_id_logs require an argument!"
```

As for the previous method, you can filter the GUID logs by passing **4** positional arguments:

* **argument**: in this case the GUID

* **since**: a start period not older than the specified date

* **to**: an end period not older than the specified date

* **lines**: the exact number of last log lines 

For example:

If you want to analyze the last 50 log lines of the GUID 1000 ranging from a month ago to 10:00 of the current day:

```ruby
Systemdy::Journal.display_group_id_logs(argument: 1000, since: '1 month ago', to: '10:00', lines: 50)
```

If you want to analyze the last 20 log lines of the GUID 1000 ranging from a week ago to yesterday:

```ruby
Systemdy::Journal.display_group_id_logs(argument: 1000, since: '1 week ago', to: 'yesterday', lines: 20)
```

You can also filter the logs by specific dates in the format YYYY-MM-DD for example:

```ruby
Systemdy::Journal.display_group_id_logs(argument: 1000, since: '2022-08-27', lines: 200)
```

If the passed arguments not match anything the method return an array with a message like this:

```ruby
["-- No entries --"]
```

### Display user_id logs (UID)

To find all messages related to a particular user, we just simply need to pass the UID of the user as argument, for example:

```ruby
Systemdy::Journal.display_user_id_logs(argument: 1000)
```

In this case the class method return an array with: 
* a default size of 10 (the last 10 lines of log)
* a start period not older than '00:00 AM' o' clock (today)
* an end period not newer than the method execution timenow

Obviously this method require the UID as argument, if executed with no arguments return a message like this:

```ruby
"display_user_id_logs require an argument!"
```

As for the previous method, you can filter the UID logs by passing **4** positional arguments:

* **argument**: in this case the UID

* **since**: a start period not older than the specified date

* **to**: an end period not older than the specified date

* **lines**: the exact number of last log lines 

For example:

If you want to analyze the last 50 log lines of the UID 1000 ranging from a month ago to 10:00 of the current day:

```ruby
Systemdy::Journal.display_user_id_logs(argument: 1000, since: '1 month ago', to: '10:00', lines: 50)
```

If you want to analyze the last 20 log lines of the UID 1000 ranging from a week ago to yesterday:

```ruby
Systemdy::Journal.display_user_id_logs(argument: 1000, since: '1 week ago', to: 'yesterday', lines: 20)
```

You can also filter the logs by specific dates in the format YYYY-MM-DD for example:

```ruby
Systemdy::Journal.display_user_id_logs(argument: 1000, since: '2022-08-27', lines: 200)
```

If the passed arguments not match anything the method return an array with a message like this:

```ruby
["-- No entries --"]
```

## Contributing

We :heart: pull requests from everyone. 

Everyone interacting in the systemdy's project codebase is expected to follow the [code of conduct](https://github.com/magic4dev/systemdy/blob/master/CODE_OF_CONDUCT.md).

### Develop a new feature

If you wanna develop a new feature:

* Fork this project

* Clone the repo with the following command:

      git clone git@github.com:your-username/systemdy.git

* Install dependencies with:

      bin/setup

* Make your changes
* Write tests. 

    **NOTE**: For a correct testing procedure read our [Note for testing](#note-for-testing)
* Make the tests pass with the following command:

      rake spec

* Add notes about your changes to the `CHANGELOG.md` file

* Write a good commit message
* Push to your fork
* [Submit a pull request](https://github.com/magic4dev/systemdy/compare)
* Wait for us, we will reply as soon as possible and if is the case we suggest some changes for a better quality of the code
* Please, if you push more than one commit let's keep the history clean :stuck_out_tongue_winking_eye:

Thank you for your contribution! :handshake:

### Fix a bug

If you wanna fix a bug:


CHANGELOG.md
## [0.4.1] - 2022-11-22

### Fixed

- Systemdy::Utility::Formatter return_an_array_from error for empty array and nil returned object

## [0.4.0] - 2022-11-22

### Added

- Systemdy::Utility::Formatter return_an_array_from class method for convert a complex string into array with custom splitter
- Systemdy::Service port method for return service port if available
- Systemdy::Service protocol method for return service protocol if available

### Changed

- Logic for Systemd::Utility::Formatter return_an_array_from_system_command class method 

## [0.3.2] - 2022-10-21 

### Fixed

- documentation_uri in gemspec

## [0.3.1] - 2022-10-21 

### Changed

- Documentation syntax for Systemdy::Utility::Formatter.remove_newline_from_system_command class method
- documentation_uri in gemspec

## [0.3.0] - 2022-10-21 

### Added

- Method remove_newline_from_system_command to Systemdy::Utility::Formatter class for remove \n characters from system calls

### Changed

- Logic for class Systemdy::Service is_active? and is_enabled? instance methods

## [0.2.1] - 2022-10-11 

### Fixed

- documentation_uri in gemspec

## [0.2.0] - 2022-10-10 

### Added

- Yard documentation of all gem's modules, classes, constants and methods(included methods generated with metaprogramming techniques)

## [0.1.0] - 2022-10-07 

### Added

- Yard-like comments for documentation of the remained gem's classes, constants and modules
- Yard-like comments for documentation of private methods

### Changed

- Changed markdown README.md structure for yard documentation

## [0.1.0] - 2022-10-06 

### Changed

- URL image path on README.md for proper yard documentation
- Comments in Systemdy::Utility::Validator class with a yard-like syntax for generate documentation
- Comments in Systemdy::Service with a yard-like syntax for generate documentation

## [0.1.0] - 2022-10-05 

### Changed

- Systemdy::Utility::Formatter comment section to yard-like syntax for generating documentation
- Variable name sudo_command to sudo in Systemdy::Service class
- Logic for the Systemdy::Utility::KeyValueFilter.filter_by_keys class method

### Added

- Yard-style comments for the Systemdy::Utility::Validator.check_if_a_service_exist class method
- Yard-style comments for the Systemdy::Utility::MessageDisplayer.display_message class method
- Yard-style comments for the Systemdy::Utility::KeyValueFilter.filter_by_keys class method

### Fixed 

- Yard params description in Systemdy::Utility::Formatter class

## [0.1.0] - 2022-10-04 

### Changed

- README.md for load image logo

### Added

- Image logo

## [0.1.0] - 2022-10-03 

### Changed

- Gem name from Systemd to Systemdy to avoid name conflicts with other rubygems with the same namespace

### Added

- Badges in README.md for render essential metadata of the project

## [0.1.0] - 2022-09-29 

### Changed

- Status method of Systemd::Service class
- Systemd::Utility::Validator.check_if_a_service_exist class method

### Added 

- Systemd::Utility::KeyValueFilter.filter_by_keys class method
- Systemd::Utility::KeyValueFilter class for filter hash with provided keys

## [0.1.0] - 2022-09-28 

### Changed

- Moved render_message method from Systemd::Utility::Validator class to Systemd::Utility::MessageDisplayer class

### Added

- Custom exception for Systemd::Utility::Formatter.return_an_array_from_system_command class method 
- Systemd::Utility::Formatter return_an_array_from_system_command class method 
- Systemd::Utility::Formatter class
- Systemd::Utility::MessageDisplayer class with render_message class method

## [0.1.0] - 2022-09-25 

### Added

- Custom test variables to TestVariables module in spec/setup/test_variables.rb with test refactor
- TestSetup module in spech_helper for share spec's variables

## [0.1.0] - 2022-09-23 

### Changed

- Systemd::Utility::Validator.exist? method with forwardable module for delegation pattern

## [0.1.0] - 2022-09-22 

### Changed

- Removed Systemd::Journal::Unit class in favour of a universal api design 

## [0.1.0] - 2022-09-19 

### Changed

- Methods is_active? and is_enabled? for return a boolean value instead of a string 

### Added

- Systemd::Journal::Unit.display_logs method for manage journalctl units
- Systemd::Journal::Unit class 

## [0.1.0] - 2022-09-18 

### Added

- Methods mask and unmask to the Systemd::Service class
- Methods is_enabled? and is_active? and rspec tests the Systemd::Service class for check if a service is enabled or active

## [0.1.0] - 2022-09-17 

### Changed 

- Values of gemspec code_of_conduct and changelog url paths

### Added

- Founded attribute to Systemd::Service class to mark a provided service if is found or not on the system
- Method exist? for check if a service exist
- Method status to Systemd::Service class with rspec test
- Actions method for execute different actions on a systemd service

## [0.1.0] - 2022-09-15

- Initial release

lib/systemdy/service.rb
module Systemdy
    # Allows to control a life-cycle of a systemd's service
    # @attr_reader command [String] the default 'systemctl' command
    # @attr_reader name [String] the name of the service
    class Service 
            
        # extend Forwardable standard's library module for delegate a specified method to a designated object
        extend Forwardable

        # service look up information command 
        INFO_LOOKUP_COMMAND           = "getent services"

        # list of essential information on a provided service
        LIST_OF_ESSENTIAL_INFO_LOOKUP = %w( port protocol )

        # list of supported actions on a provided service
        LIST_OF_ACTIONS               = %w( start restart stop enable disable reload mask unmask )

        # list of supported statuses on a provided service
        LIST_OF_STATUSES              = %w( enabled active )

        # list of status properties on a provided service when status command is called
        LIST_OF_STATUS_PROPERTIES     = %w( Id Description ExecMainPID LoadState ActiveState FragmentPath 
                                            ActiveEnterTimestamp InactiveEnterTimestamp ActiveExitTimestamp 
                                            InactiveExitTimestamp 
                                        )

        attr_reader :command, :name

        # method for create a new Systemdy::Service object 
        #
        # @param  name [String] the name of the systemd service to control
        # @return [Systemdy::Service] a new Systemdy::Service object
        # @example Create an object
        #    my_postgresql_service = Systemdy::Service.new('postgresql')
        def initialize(name)
            @command   = SYSTEMCTL_COMMAND # constant contained in Systemdy.rb
            @name      = name 
        end

        # delegate return_an_array_from_system_command method to Systemdy::Utility::Formatter class contained in Systemdy/utility/formatter.rb
        def_delegator Systemdy::Utility::Formatter,        :return_an_array_from_system_command
        # delegate return_an_array_from method to Systemdy::Utility::Formatter class contained in Systemdy/utility/formatter.rb
        def_delegator Systemdy::Utility::Formatter,        :return_an_array_from
        # delegate remove_newline_from_system_command to Systemdy::Utility::Formatter class contained in Systemdy/utility/formatter.rb
        def_delegator Systemdy::Utility::Formatter,        :remove_newline_from_system_command
        # delegate render_message method to Systemdy::Utility::MessageDisplayer class contained in Systemdy/utility/message_displayer.rb
        def_delegator Systemdy::Utility::MessageDisplayer, :render_message 
        # delegate check_if_a_service_exist method to Systemdy::Utility::Validator class contained in Systemdy/utility/validator.rb
        def_delegator Systemdy::Utility::Validator,        :check_if_a_service_exist 
        # delegate filter_by_keys method to Systemdy::Utility::KeyValueFilter class contained in Systemdy/utility/key_value.rb
        def_delegator Systemdy::Utility::KeyValueFilter,   :filter_by_keys 

        # method for check if a created service +exist+
        #
        # @return [Boolean] the presence of the service on the system
        # @example Create an object
        #   my_postgresql_service.exist? #=> true
        def exist? 
            check_if_a_service_exist(name) # class method contained in Systemdy/utility/validator.rb
        end

        # create dynamically methods based on LIST_OF_ESSENTIAL_INFO_LOOKUP constant 
        # @!method port
        #   return the available +port+ of the service
        #   @example get the service port
        #       my_postgresql_service.port #=> "5432"
        #   @note This method is generated with use of metaprogramming techniques
        #   @todo This method return an error message when there are no port available
        #
        # @!method protocol
        #   return the available +protocol+ of the service
        #   @example restart a service
        #       my_postgresql_service.protocol #=> "tcp"
        #   @note This method is generated with use of metaprogramming techniques
        #   @todo This method return an error message when there are no protocol available
        #
        LIST_OF_ESSENTIAL_INFO_LOOKUP.each_with_index do |info, index|
            define_method info do 
                return default_error_message() unless exist?
                essential_info = return_an_array_from(`#{INFO_LOOKUP_COMMAND} #{name}`, argument_splitter: ' ')
                return info_lookup_error_message(info) if essential_info.nil? || essential_info.empty?
                return_an_array_from(essential_info[1], argument_splitter: '/')[index]
            end
        end
        
        # create dynamically methods based on LIST_OF_ACTIONS constant
        # @!method start
        #   execute action +start+ on the service
        #   @example start a service
        #       my_postgresql_service.start
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method restart
        #   execute action +restart+ on the service
        #   @example restart a service
        #       my_postgresql_service.restart
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method stop
        #   execute action +stop+ on the service
        #   @example stop a service
        #       my_postgresql_service.stop
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method enable
        #   execute action +anable+ on the service
        #   @example enable a service
        #       my_postgresql_service.enable
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method disable
        #   execute action +disable+ on the service
        #   @example disable a service
        #       my_postgresql_service.disable
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method reload
        #   execute action +reload+ on the service
        #   @example reload a service
        #       my_postgresql_service.reload
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method mask
        #   execute action +mask+ on the service
        #   @example mask a service
        #       my_postgresql_service.mask
        #   @note This method is generated with use of metaprogramming techniques
        #
        # @!method unmask
        #   execute action +unmask+ on the service
        #   @example unmask a service
        #       my_postgresql_service.unmask
        #   @note This method is generated with use of metaprogramming techniques
        #
        LIST_OF_ACTIONS.each do |action|
            define_method action do 
                return default_error_message() unless exist?
                sudo = Etc.getpwuid(Process.uid).name != 'root' ? 'sudo' : ''
                `#{sudo} #{command} #{action} #{name}` 
            end
        end

        # method for return a key/value pair of the service's properties
        #
        # @return [Hash] the service's properties
        # @example display all the properties related to a service
        #   
        #   my_postgresql_service.properties 
        #
        #     {
        #       "Type"=>"oneshot",              
        #       "Restart"=>"no",          
        #       "ExecMainPID"=>"48615",
        #       "NotifyAccess"=>"none"
        #     }
        #
        # @note For the sake of brevity, all service-related properties are not shown in this example
        def properties
            array_of_properties = return_an_array_from_system_command(`#{command} show #{name}`)
            array_of_properties.collect { |property| { property.split('=')[0] => property.split('=')[1] } }.reduce({}, :merge)
        end

        # method for return the current status of the provided service
        #
        # @return [Hash] the service's current status
        # @example display the current status of the service
        #   my_postgresql_service.status
        #
        #     {
        #       "Id"=>"postgresql.service",              
        #       "Description"=>"PostgreSQL RDBMS",          
        #       "ExecMainPID"=>"48615",
        #       "LoadState"=>"loaded",
        #       "ActiveState"=>"active",
        #       "FragmentPath"=>"/lib/Systemdy/system/postgresql.service",
        #       "ActiveEnterTimestamp"=>"Thu 2022-09-29 17:13:07 CEST",
        #       "InactiveEnterTimestamp"=>"Thu 2022-09-29 17:12:44 CEST",
        #       "ActiveExitTimestamp"=>"Thu 2022-09-29 17:12:44 CEST",
        #       "InactiveExitTimestamp"=>"Thu 2022-09-29 17:13:07 CEST"
        #     }
        #
        def status 
            return default_error_message() unless exist?
            filter_by_keys(properties, LIST_OF_STATUS_PROPERTIES) 
        end

        # create dynamically methods based on LIST_OF_STATUSES constant
        # @!method is_enabled?
        #   check if the service is +enabled+
        #   @return [Boolean] the service is enabled
        #   @example check if a service is enabled
        #       my_postgresql_service.is_enabled? #=> true
        # @note This method is generated with use of metaprogramming techniques
        #
        # @!method is_active?
        #   check if the service is +active+
        #   @return [Boolean] the service is active
        #   @example check if a service is active
        #       my_postgresql_service.is_active? #=> true
        # @note This method is generated with use of metaprogramming techniques
        LIST_OF_STATUSES.each do |status|
            define_method "is_#{status}?" do 
                return default_error_message() unless exist?
                remove_newline_from_system_command(`#{command} is-#{status} #{name}`) == status
            end
        end

        # method to show a message when an action is performed on a service not installed on the system
        # @return [String] the error message
        # @example the action start on a service not installed on the system
        #     a_service_not_installed.start #=> "Unit a_service_not_installed.service could not be found."
        def default_error_message
            render_message("Unit #{name}.service could not be found.") # class method contained in Systemdy/utility/message_displayer.rb
        end

        # method to show a message when a service has no port or protocols available
        # @return [String] the info lookup error message
        # @example call port method on a service that has no port available
        #     a_service_that_has_no_port_available #=> "a_service_that_has_no_port_available.service has no port available"
        def info_lookup_error_message(info)
            render_message("#{name}.service has no #{info} available") # class method contained in Systemdy/utility/message_displayer.rb
        end

        # @!method return_an_array_from_system_command
        #   @param system_call [String] system call to convert to an array
        #   @return [Array] an array-based list of the values ​​returned by making a system call
        #   @note check out more about this method in Systemdy/utility/formatter.rb
        #
        # @!method return_an_array_from
        #   @param string_to_parse [String] string to convert to an array
        #   @param argument_splitter [String] character for split string to an array 
        #   @param remove_blank_elements [Boolean] remove blank elements or not 
        #   @return [Array] an array-based list of the values ​​returned by argument_splitter
        #
        # @!method filter_by_keys
        #   @param hash [Hash] the hash to filter
        #   @param list_of_keys [Array] the list of keys to extract from the hash
        #   @return [Hash] a new hash that contains only the required keys
        #   @note check out more about this method in Systemdy/utility/key_value_filter.rb
        #

        # make the methods below as private
        private :render_message, :default_error_message, :info_lookup_error_message, :return_an_array_from_system_command, :return_an_array_from, :filter_by_keys 
    end
end